deps/undici/src/lib/observable.js

```diff
@@ -0,0 +1,802 @@
+            }
+            hasState = true;
+          },
+          error: reject,
+          complete() {
+            resolve(state);
+          },
+        },
+        {
+          signal,
+        }
+      );
+    });
+  }
+
+  toArray(options) {
+    return this.reduce(
+      (arr, value) => {
+        arr.push(value);
+        return arr;
+      },
+      [],
+      options
+    );
+  }
+
+  catch(handleError) {
+    return new Observable((destination) => {
+      this.subscribe(
+        {
+          next(value) {
+            destination.next(value);
+          },
+          error(error) {
+            Observable.from(handleError(error)).subscribe(destination, {
+              signal: destination.signal,
+            });
+          },
+          complete() {
+            destination.complete();
+          },
+        },
+        {
+          signal: destination.signal,
+        }
+      );
+    });
+  }
+
+  finally(onFinalize) {
+    return new Observable((destination) => {
+      destination.addTeardown(onFinalize);
+      this.subscribe(destination, {
+        signal: destination.signal,
+      });
+    });
+  }
+
+  switchMap(project) {
+    return new Observable((destination) => {
+      let index = 0;
+      let outerComplete = false;
+      let innerAC;
+      this.subscribe(
+        {
+          next: (value) => {
+            innerAC?.abort();
+            innerAC = new AbortController();
+            const signal = AbortSignal.all([
+              innerAC.signal,
+              destination.signal,
+            ]);
+            let innerObservable;
+            try {
+              innerObservable = Observable.from(project(value, index++));
+            } catch (error) {
+              destination.error(error);
+              return;
+            }
+            innerObservable.subscribe(
+              {
+                next(innerValue) {
+                  destination.next(innerValue);
+                },
+                error(error) {
+                  destination.error(error);
+                },
+                complete() {
+                  innerAC = undefined;
+                  if (outerComplete) {
+                    destination.complete();
+                  }
+                },
+              },
+              {
+                signal,
+              }
+            );
+          },
+          error(error) {
+            destination.error(error);
+          },
+          complete() {
+            outerComplete = true;
+            if (!innerAC) {
+              destination.complete();
+            }
+          },
+        },
+        {
+          signal: destination.signal,
+        }
+      );
+    });
+  }
+
+  do(fnOrObserver) {
+    return new Observable((destination) => {
+      const doObserver =
+        typeof fnOrObserver === 'function'
+          ? { next: fnOrObserver }
+          : fnOrObserver;
+      this.subscribe(
+        {
+          next(value) {
+            doObserver.next?.(value);
```

Comment:
Should be more performant to create a local `next` variable here, instead of an object `{ next: fnOrObserver }`:

```suggestion
      const next = 
        typeof fnOrObserver === 'function'
          ? fnOrObserver
          : fnOrObserver.next;
      this.subscribe(
        {
          next(value) {
            next?.(value);
```